core/src/ast/utils.rs

```diff
@@ -21,3 +24,224 @@ pub fn get_doc_lines(attrs: &[Attribute]) -> String {
 
     lines
 }
+
+#[derive(Clone, PartialEq, Eq, Hash, Serialize, Deserialize, Debug)]
+pub struct RustLink {
+    path: Path,
+    typ: DocType,
+}
+
+#[derive(Clone, PartialEq, Eq, Hash, Serialize, Deserialize, Debug)]
+enum DocType {
+    Struct,
+    StructField,
+    Enum,
+    EnumVariant,
+    EnumVariantField,
+    Trait,
+    FnInStruct,
+    FnInEnum,
+    FnInTrait,
+    DefaultFnInTrait,
+    Fn,
+    Mod,
+    Constant,
+    Macro,
+}
+
+pub fn get_rust_link(attrs: &[Attribute]) -> Option<RustLink> {
+    for attr in attrs {
+        if attr.path.to_token_stream().to_string() == "diplomat :: rust_link" {
+            if let Ok(syn::Meta::List(syn::MetaList { nested, .. })) = attr.parse_meta() {
+                if nested.len() == 2 {
+                    if let (
+                        syn::NestedMeta::Meta(syn::Meta::Path(path)),
+                        syn::NestedMeta::Meta(syn::Meta::Path(typ)),
+                    ) = (&nested[0], &nested[1])
+                    {
+                        if let Some(typ) = typ.get_ident() {
+                            return Some(RustLink {
+                                path: Path::from_syn(path),
+                                typ: match typ.to_token_stream().to_string().as_str() {
+                                    "Struct" => DocType::Struct,
+                                    "StructField" => DocType::StructField,
+                                    "Enum" => DocType::Enum,
+                                    "EnumVariant" => DocType::EnumVariant,
+                                    "EnumVariantField" => DocType::EnumVariantField,
+                                    "Trait" => DocType::Trait,
+                                    "FnInStruct" => DocType::FnInStruct,
+                                    "FnInEnum" => DocType::FnInEnum,
+                                    "FnInTrait" => DocType::FnInTrait,
+                                    "DefaultFnInTrait" => DocType::DefaultFnInTrait,
+                                    "Fn" => DocType::Fn,
+                                    "Mod" => DocType::Mod,
+                                    "Constant" => DocType::Constant,
+                                    "Macro" => DocType::Macro,
+                                    x => panic!("Invalid doc type {:?}", x),
+                                },
+                            });
+                        }
+                    }
+                }
+            }
+            panic!("Malformed attribute: {}", attr.to_token_stream());
+        }
+    }
+
+    None
+}
+
+impl RustLink {
+    pub fn http(&self, base_urls: &HashMap<String, String>) -> String {
```

Comment:
Suggestion: I think a better way to structure this would be

```rust
struct DocsUrlGenerator {
    base_urls: HashMap<String, String>
}

impl DocsUrlGenerator {
    pub fn gen_for_rust_link(rust_link: &RustLink) { ... }
}
```

Reply:
+1

Though I think we should perhaps have a "default" alongside base_urls, with the "default" defaulting to docs.rs. Most users will probably want to either have docs.rs default or point to something like https://unicode-org.github.io/icu4x-docs/doc/icu/index.html

Reply:
Done, except for the docs.rs default, which is tricky as it's a different ULR for every crate.

Reply:
I guess from my POV ICU4X will be running this for the unicode-org crates and not linking to anything but the unicode-org crates; so it would be nice to change the default from "docs.rs with magic" to "unicode-org link"

We could support `{}`-replacement in the default URL, I guess.

Reply:
I'm fine landing this as is and iterating on the UI exposed to users later htough